File: backend/src/main/java/com/snackoverflow/toolgether/domain/Notification/NotificationService.java
package com.snackoverflow.toolgether.domain.Notification;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.snackoverflow.toolgether.domain.user.service.UserService;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class NotificationService {

	private final NotificationRepository notificationRepository;
	private final UserService userService;
	private final ApplicationEventPublisher publisher;

	// 알림 생성 및 저장
	@Transactional
	public Notification createNotification(Long userId, String message) {
		Notification notification = Notification.builder()
			.user(userService.findUserById(userId))
			.message(message)
			.createdAt(LocalDateTime.now())
			.isRead(false)
			.build();
		Notification savedNotification = notificationRepository.save(notification);

		// 이벤트 발행
		publisher.publishEvent(new NotificationCreatedEvent(this, userId, message));
		return savedNotification;
	}

	// 사용자 ID로 읽지 않은 알림 목록 조회
	@Transactional(readOnly = true)
	public List<NotificationDto> getUnreadNotifications(Long userId) {
		return notificationRepository.findByUserIdAndIsRead(userId, false)
			.stream()
			.map(NotificationDto::new)
			.collect(Collectors.toList());
	}

	@Transactional
	public void markNotificationAsRead(Long notificationId) {
		Optional<Notification> notification = notificationRepository.findById(notificationId);
		if (notification.isPresent()) {
			Notification n = notification.get();
			n.setIsRead(true);
			notificationRepository.save(n);
		} else {
			throw new RuntimeException("Notification not found");
		}
	}
}

File: backend/src/main/java/com/snackoverflow/toolgether/domain/Notification/service/NotificationService.kt
package com.snackoverflow.toolgether.domain.notification.service

import com.snackoverflow.toolgether.domain.notification.NotificationCreatedEvent
import com.snackoverflow.toolgether.domain.notification.dto.NotificationDto
import com.snackoverflow.toolgether.domain.notification.entity.Notification
import com.snackoverflow.toolgether.domain.notification.repository.NotificationRepository
import com.snackoverflow.toolgether.domain.user.service.UserService
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.*
import java.util.stream.Collectors

@Service
class NotificationService(
    private val notificationRepository: NotificationRepository,
    private val userService: UserService,
    private val publisher: ApplicationEventPublisher
) {
    // 알림 생성 및 저장
    @Transactional
    fun createNotification(userId: Long?, message: String): Notification {
        val notification: Notification = Notification(
            userService.findUserById(userId),
            message,
            LocalDateTime.now(),
            false
        )
        val savedNotification: Notification = notificationRepository.save(notification)

        // 이벤트 발행
        publisher.publishEvent(NotificationCreatedEvent(this,  userId ?: -1L, message))
        return savedNotification
    }

    // 사용자 ID로 읽지 않은 알림 목록 조회
    @Transactional(readOnly = true)
    fun getUnreadNotifications(userId: Long): List<NotificationDto> {
        return notificationRepository.findByUserIdAndIsRead(userId, false)
            .stream()
            .map { notification: Notification? ->
                NotificationDto(
                    notification!!
                )
            }
            .collect(Collectors.toList())
    }

    @Transactional
    fun markNotificationAsRead(notificationId: Long) {
        val notification: Optional<Notification?> = notificationRepository.findById(notificationId)
        if (notification.isPresent) {
            val n: Notification = notification.get()
            n.setIsRead(true)
            notificationRepository.save<Notification>(n)
        } else {
            throw RuntimeException("Notification not found")
        }
    }
}